프로젝트 매니저 [STEP 2] 하모, Wonbi

wonbi92 opened the pull request:
@junbangg 
알라딘 행복한 명절 지내셨나요??
드디어 STEP2의 PR을 보내게 되었습니다!

## 📝 STEP 진행 중 경험하고 배운 것
- `MVVM`
- `책임주도설계`
- `Container ViewController`
- `PopOverViewController`

## 💭 고민한 부분

## 모델의 state를 어떻게 구분지을까
- enum 사용
    - ocp를 위반하는 경우가 존재한다.
    - case가 추가될 경우 기존의 코드를 수정해야 하는 상황이 발생한다.
- 모델 프로토콜을 이용하여 각 state별 구체 타입 생성
    - state가 추가되더라고 기존의 코드를 수정하지 않고 확장할 수 있음
    - 복잡성이 증가한다.
 
현재 프로젝트에서 핵심 로직 중 하나는 todo, doing, done 세가지 종류의 데이터를 분류하는 로직이라 생각했습니다.
이 로직을 수행하기 위해 모델 안에 enum을 사용해서 각 케이스를 만들고, 그 케이스별 분기처리를 하는것이 맞는것인가? 아니면 세가지의 각각의 다른 모델을 만들고 그 모델에 해당하는 ViewModel과 View를 각각 처리하는 방법이 맞는것인가? 라는 고민을 했습니다.
또 아예 방향성이 잘못된 것은 아닌가? 세가지 상태를 꼭 분류할 필요가 있는가? 상태라는 하나의 값 때문에 세가지의 모델이 생기는 것이 옳은 방향성인가? 상태는 모델의 한가지의 속성에 불과한데, 이 프로퍼티 하나 때문에 모델이 분기 되는 것이 맞는 것인가? 라는 고민에 도달하게 되었습니다.

결론적으로 각 상태에 따라 모델을 가지는 것이 아니라 하나의 모델이 상태를 열거형으로 가지고 있는 방향으로 사용하게 되었습니다. OCP를 위반하는 것이 아닐까 라는 생각이 들었지만, 그 리스크보다 훨씬 더 좋은 코드 가독성과 데이터를 처리하는 로직을 단순화 할 수 있는 리턴이 더 크다고 결론짓고 이 방식으로 사용하게 되었습니다.


## Date의 처리
Date타입의 데이터를 모델이 가지고 있습니다. 모델은 Date타입으로 Deadline을 가지고 있지만, 뷰는 이를 String으로 받아서 표현해야합니다. 이 값을 트렌스폼 하는 객체는 누가 되야 하는지 고민하였습니다. 결과적으로 Cell의 ViewModel을 만들어 Date타입의 값을 String으로 변경하는 로직을 가지게 하고 Date Extension을 통해서 DateFormatter와 formatting메서드를 만들어서 사용하였습니다.

## MainViewModel에서 각 테이블뷰에 뿌릴 모델을 왜 쪼갰는가
하나의 배열로 관리하게 되면 변동사항이 하나 생겼을 때 모든 테이블뷰에 데이터를 바꾸는 작업을 해야하기 때문에 3개로 나누어서 변경이 발생하는 테이블뷰만 작업을 수행하도록 하려고 했습니다.
이렇게 하면 각 테이블뷰가 자신의 모델만 바라보게 되고, 모든 테이블 뷰가 데이터를 바꾸는 작업을 하지 않게 되는 장점이 있었습니다. 하지만 분기처리가 많이 생기는 단점이 발생했습니다.

## 🤔 조언을 얻고 싶은 부분
### MainViewModel에서 각 테이블뷰에 뿌릴 모델을 쪼개면서 발생한 분기처리
메인 뷰모델에서 모델을 3개 가지게 되면서, project의 state에 따라 분기처리를 해줘야 하는 문제가 발생하게 되었습니다. 특히 업데이트를 하는 경우, 해당 데이터의 id를 통해서 인덱스를 가져와 그 인덱스에 해당하는 데이터를 교체해주는 로직을 사용하고 있는데 이 로직이 guard문이 필요해 switch문 안에 guard문이 들어가는 구조가 되었습니다. 
그래서 로직을 분리하다 보니 inout을 쓰게 되는 등의 문제가 생기게 되었습니다. 이러한 방향성이 맞는 방향성인지 궁금합니다.

### ViewModel끼리 Delegate를 통해서 이벤트전달
프로젝트를 편집할 때 MainViewModel에 있는 동일한 id값을 가진 요소를 수정해야해서 id값이 필요했습니다. 그런데 EditViewController의 경우 id를 가지고있을 위치가 없다고 생각해서 EditViewModel을 만들어서 변경이 완료되었을 때 id값과 함께 delegate를 통해 MainViewModel로 이벤트를 전달하였습니다. 이런 구조에 대해서 알라딘의 의견이 궁금합니다!!

wonbi92 committed:
build: FireBase, SwiftLint 추가 및 적용

lxodud committed:
feat: UITableView 3개 구현

- 스토리보드 삭제
- window rootView 설정
- view autoLayout 설정

wonbi92 committed:
feat: 각 테이블 뷰에 해당하는 Header 커스텀 뷰 구현

- 각 테이블 뷰의 셀의 갯수를 나타내는 커스텀 레이블 구현
- 각 헤더를 담는 스택뷰 구현

wonbi92 committed:
refactor: 셀의 갯수를 나타내는 레이블 크기 조정

- 가로가 세로보다 크거나 같도록 제약조건 추가
- 숫자가 원 중앙에 위치하도록 얼라이먼트 추가

lxodud committed:
refactor: CountLabel, HeaderView final 키워드 추가

- CountLabel 클래스 padding 프로퍼티 초기값 지정

lxodud committed:
docs: README 수정

wonbi92 committed:
docs: README 수정

lxodud committed:
style: 파일 계층 구조 수정

wonbi92 committed:
feat: MainViewModel, ProjectModel 객체 생성

- DIP를 위한 Projectable, MainViewModelProtocol 구현
- 각각의 프로토콜을 채택하는 콘크리트 객체 생성
- 소스파일 주석 정리

lxodud committed:
feat: ProjectCell 타입 구현

wonbi92 committed:
refactor: 각각의 테이블 뷰를 컨트롤 하는 ProjectListViewController 구현

- 기존 MainViewController의 자식으로 ProjectListViewController를 가지도록 리팩토링
- MainViewController 메서드 정리

lxodud committed:
feat: ProjectListViewController tableView diffable datasource 구성

- 모델 타입 Hashable 채택
- ProjectListViewController의 재사용성을 위한 제네릭 타입 사용

wonbi92 committed:
feat: 프로젝트를 추가하거나 수정할 때 나타나는 EditorViewController 구현

- EditorViewController가 가지고 있는 요소의 속성 설정
- MainViewController의 네비게이션 바의 Add버튼 및 버튼 액션 추가
- ProjectListViewController의 제너릭 타입명을 ModelType으로 변경

lxodud committed:
feat: EditorViewController 네비게이션 버튼 구현

wonbi92 committed:
refactor: 모델이 state별로 나뉘지 않고, state를 프로퍼티로 가지도록 리팩토링

- Project모델 타입 재구성
- State enum타입 구현
- 모델 타입 재구성으로 생긴 사이드 이팩트 정리

lxodud committed:
feat: EditorViewController Done 버튼 액션 구현

- MainViewModel을 통해서 Project를 생성
- MainViewController에서 MainViewModel 바인딩 구현

wonbi92 committed:
feat: ProjectListViewController의 테이블뷰 Delegate 및 cell삭제 로직 구현

- ProjectListViewController 로직 분리
- viewModel의 deleteProject메서드 로직 구현

lxodud committed:
feat: ProjectListActionDelegate 구현

- delete 로직을 delegate가 하도록 수정

wonbi92 committed:
feat: cell을 누르면 Editor뷰를 띄우는 로직 구현

- EditorViewController의 convenience 이니셜라이저 구현
- 오타 수정
- editProject메서드 로직 구현

lxodud committed:
refactor: 생성, 편집뷰 EditViewController, AddViewController로 분리

- 생성, 편집뷰의 공통된 뷰를 가지고있는 ProjectViewController를 상속

wonbi92 committed:
feat: EditViewController 로직 구현

- EditViewModel 구체 생성
- ProjectViewController 프로퍼티 접근 제어 수정

lxodud committed:
feat: EditViewController userInteraction 핸들링 구현

- EditViewModel editMode 상태 로직 구현
- AddViewController, EditViewController 네비게이션 바 구성

wonbi92 committed:
feat: EditViewController의 데이터 바인딩 구현

- AddViewController의 Delegate 구현
- AddActionDelegate를 MainViewController로 지정
- EditViewController의 ViewProject 모델 구현
- Model의 deadline타입을 Calendar로 변경

lxodud committed:
feat: 프로젝트 편집 기능 구현

- EditProjectDelegate을 통해서 EditViewModel에서 MainViewModel로 데이터 전달

wonbi92 committed:
refactor: EditProjectDelegate대신 AddViewModel이 데이터를 넘겨주도록 리팩토링

- 이제 AddViewModel과 EditViewModel은 ViewModelDelegate를 대리자로 가집니다.
- AddViewController의 이니셜라이저에서 AddViewModel을 주입받도록 구현

lxodud committed:
feat: ProjectCellViewModel 구현

- ProjectCell이 ProjectCellViewModel을 바인딩하여 데이터 관련 로직 분리

wonbi92 committed:
feat: 테이블뷰의 셀의 갯수와 헤더의 타이틀을 가지는 HeaderViewModel구현

- HeaderView의 countLabel이 cell의 갯수가 바뀔 때 마다 업데이트 되는 로직 구현

lxodud committed:
feat: ProjectCell LongPressGesture 추가

wonbi92 committed:
refactor: MainViewModel이 하나의 모델을 가지는 것이 아니라 각 state에 맞는 모델 배열을 3개 가지도록 리팩토링

- MainViewModelProtocol 삭제

lxodud committed:
feat: Cell의 state 변경 액션 추가

- State 타입에 CaseIterable 채택
- MainViewModel changeProjectState 메서드 추가

wonbi92 committed:
design: 셀의 폰트와 모양 변경

lxodud committed:
feat: ProjectCell DateLabel Color 변경 로직 구현

- DateFormatter를 Date 타입의 타입 프로퍼티로 구현
- Project모델의 deadline 타입 변경

wonbi92 committed:
feat: AddViewController, EditViewController 타이틀 구현

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/HeaderView.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    private func bindHeaderViewModel() {
+        headerViewModel.cellCountHandler = { [weak self] count in
+            self?.countLabel.text = count
+        }
+        
+        headerViewModel.titleHandler = { [weak self] title in
+            self?.titleLabel.text = title
+        }
```
handler 라는 이름은 어쩌다 붙게 된건가요??? (틀렸다는건 아닙니다)

wonbi92 replied:
클로저를 변수로 사용하면서 통상적으로 Handler를 많이 사용한다 생각했고, 뷰 모델에 있는 상태가 변경될 때 수행할 로직을 담고 있다 생각해서 핸들러라는 이름과 잘 어울리는거 같아 쓰게 되었습니다.

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/HeaderView.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    private func bindHeaderViewModel() {
+        headerViewModel.cellCountHandler = { [weak self] count in
+            self?.countLabel.text = count
+        }
+        
+        headerViewModel.titleHandler = { [weak self] title in
+            self?.titleLabel.text = title
+        }
```
self 언래핑 해주는건 어떨까요?!

wonbi92 replied:
self를 언래핑해서 강한 참조를 시키면 이 self가 지연 해제 되는 걸로 알고 있습니다. 결론적으로 self 언래핑을 할지 안할지는 내부 로직에 발생하는 비용이나 다른 참조들을 고려해서 결정해야 한다 생각합니다. 
다만, 이코드에서는 하나의 프로퍼티에 값을 할당하는 간단한 로직만 있기 때문에 언래핑을 하던 안하던 큰 차이가 없다고 생각했습니다. 혹시 언래핑을 해줘야 하는 저희가 생각하지 못한 이유가 있을까요?

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/ProjectCell.swift:
```diff
@@ -0,0 +1,121 @@
+}
+
+final class ProjectCell: UITableViewCell {
+    static let reuseIdentifier = String(describing: ProjectCell.self)
```
흠 그냥 `static let identifier = "ProjectCell"` 로 안하고 `String(describing: ProjectCell.self)` 를 사용하시게 된 이유가 궁금합니다~

lxodud replied:
Cell 타입이 많아졌을 때 Cell의 reuseIdentifier를 리터럴값으로 하게되면 name space를 만들어서 관리를 해주어야하는데 init(describing:)를 통해 자기 자신의 클래스이름을 reuseIdentifier로 사용하여 따로 관리하지 않아도 된다고 생각했습니다.
더 나아가서 UITableViewCell의 extension이나 Reusable이라는 프로토콜을 통해서 computed property를 구현해주면 보일러플레이트 코드도 줄일 수 있다고 생각합니다!!

junbangg replied:
그렇군요!
그렇다면 지금 해주신 `String(describing: ProjectCell.self)` 을 살짝 응용한 프로토콜로도 만들어서 사용할 수 있습니다.

```
protocol IdentifiableView {
    static var identifier: String { get }
}

extension IdentifiableView {
    static var identifier: String {
        return String(describing: self)
    }
}
```

이렇게 해서 identifier이 필요한 모든 타입에 채택시켜서
`ProjectCell.identifier` 식으로 똑같이 쓸 수 있어요!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/ProjectCell.swift:
```diff
@@ -0,0 +1,121 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    private func configureSubViews() {
```
`SubView` 는 단어가 하나라서 그냥 `Subview` 로 가도 될것 같습니다!

wonbi92 replied:
넵 수정했습니다!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/ProjectCell.swift:
```diff
@@ -0,0 +1,121 @@
+
+final class ProjectCell: UITableViewCell {
+    static let reuseIdentifier = String(describing: ProjectCell.self)
+    private var projectCellViewModel: ProjectCellViewModel?
```
`viewModel` 이 여러개가 있는것도 아닌 상황이고, 옆에 명시되어있는 타입도 가독성에 어느정도 기여를 하기 때문에 그냥 `viewModel` 로 가도 될것 같네요! (개인적인 생각쓰)

wonbi92 replied:
넵 수정했습니다!

lxodud replied:
저도 동의합니다 알라딘 ㅎㅎㅎ

junbangg commented on ProjectManager/ProjectManager/Sources/View/Components/ProjectCell.swift:
```diff
@@ -0,0 +1,121 @@
+// MARK: Configure ProjectListCellViewModel
+extension ProjectCell {
+    func setupViewModel(_ viewModel: ProjectCellViewModel) {
+        projectCellViewModel = viewModel
```
만약 위에서 저의 `projectCellViewModel` -> `viewModel` 의견을 받아주신다면
여기서 `self` 도 붙여줘야겠네요 :)

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/AddViewController.swift:
```diff
@@ -0,0 +1,58 @@
+import UIKit
+
+final class AddViewController: ProjectViewController {
+    private let viewModel: AddViewModel?
```
오 여기서는 그냥 `viewModel` 로 가셨군요!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/AddViewController.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+    
+    private func tapDoneButton(_ sender: UIAction) {
+        guard let title = textField.text,
+              let description = textView.text
+        else {
+            return
+        }
```
그냥 이런식으로 하는게 좀 더 자연스러워 보이네요 저는!
```suggestion
        guard let title = textField.text,
              let description = textView.text else {
              return
        }
```

wonbi92 replied:
저희가 프로젝트 시작 전에 코드 스타일에 대해서 정리를 한번 하고 시작했었는데요! 그 때 가드문에 여러 조건이 있으면 else 문을 개행하는 것을 규칙으로 삼아서 요렇게 진행하였습니다!

wonbi92 replied:
알라딘이 제시해주신 부분도 충분히 좋은 방법인것 같습니다!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/EditViewController.swift:
```diff
@@ -0,0 +1,85 @@
+            self?.textField.text = viewProject.title
+            self?.datePicker.date = viewProject.deadline
+            self?.textView.text = viewProject.description
+        }
```
흠 `viewProject` 를 통째로 전달해서 이렇게 사용하는것보다
`viewModel` 단에서 미리 `textField`, `datePicker.date`, `textView.text` 의 내용을 미리 정해서 따로따로 바인딩하는건 어떨까요?? 확장성을 생각했을때 그게 좋을것 같다는 저의 개인적인 생각입니다!

junbangg replied:
또한, `Project` 라는 타입이 여기 `View` 까지 따라와서 의존을 굳이 할 필요가 없어보입니다. `ViewModel` 이 있는데 `Project` 라는 모델을 `View` 에서 알 필요가 없어야할것 같아요!
`ViewModel` 에서 처리하고 분해해서 `View` 가 소화할 수 있는 형태로 미리 준비해주는게 좋아보입니다!

만약에 고런게 고려 되었었고 이렇게 하신 다른 이유가 있었다면 말씀해주십셔!

lxodud replied:
해당 project를 수정하는 흐름이다 보니까 project의 id가 필요했고 그래서 뷰모델에 Project 타입의 프로퍼티를 만들어서 수정할 데이터를 해당 프로퍼티에 저장하는 김에 이 프로퍼티에 바인딩을 했던 것 같습니다. 생각해보니 View가 Project를 전부 알 필요가 없기 때문에 뷰에서 필요한 값들만 넘겨줄 수 있게 구현하는게 맞는 것 같습니다!!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,135 @@
+import UIKit
+
+final class MainViewController: UIViewController {
+    private let toDoViewController = ProjectListViewController(state: .todo)
+    private let doingViewController = ProjectListViewController(state: .doing)
+    private let doneViewController = ProjectListViewController(state: .done)
```
요 친구들도 주입 시키는 방법은 어떨까요??

lxodud replied:
요 친구들을 주입 시켰을 때 기본값을 바로 지정해주는 것 보다 더 유연한 클래스가 되고 프로토콜로 추상화까지 한다면 훨씬 더 유연한 클래스가 될 것 같습니다. 알라딘이 이 코드에서 주입 시키는 방법을 떠올린 이유와 동일할까요???

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,135 @@
+
+// MARK: ProjectListAction Delegate
+extension MainViewController: ProjectListActionDelegate {
+    func changeProjectState(with state: State, to project: Project) {
+        viewModel.changeStateProject(state: state, project: project)
+    }
```
(제가 코드를 100% 이해 안하고 하는 이야기 일 수도 있어서 주의!)

`ProjectListActionDelegate` 를 통해서 `ViewController` 를 거쳐서 `ViewModel` 한테 `changeStateProject` 시키는것 말고
그냥 `ProjectListActionDelegate` 를 `viewModel` 에 직접 채택 시켜서 바로 `viewModel` 시키는건 불가능한 구조일까요?

밑에 `delete` 도 그렇고요!

junbangg replied:
지금도 View가 아주 깡통이 됐지만 
조금이라도 더 깡통을 만들 수 없을까 하고 의견 드려보는겁니다! 될지는 잘 모르겠어요 저도! ㅎㅎ

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/MainViewController.swift:
```diff
@@ -0,0 +1,135 @@
+        viewModel.changeStateProject(state: state, project: project)
+    }
+    
+    private func configureProjectListActionDelegate() {
+        [toDoViewController, doingViewController, doneViewController].forEach {
+            $0.delegate = self
+        }
+    }
```
오 이 친구는 `ProjectListActionDelegate` 메서드가 아니네요???
이건 위치를 이동시키는게 좋을것 같네요🤔 혼란을 줄 여지가 다소 있어보입니다~!

wonbi92 replied:
넵 이동시키는걸 깜빡했네요 😅 수정하였습니다!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/ProjectListViewController.swift:
```diff
@@ -0,0 +1,195 @@
+    private let tableView = UITableView()
+    private var dataSource: UITableViewDiffableDataSource<Section, Project>?
+    private let state: State
+    weak var delegate: ProjectListActionDelegate?
```
다른 곳들도 더 많을것 같지만 여기에만 댓글 남깁니다ㅎㅎ  `delegate` 는 은닉을 시킬 수 없는걸까요?

wonbi92 replied:
private 접근제어자를 통해 은닉화를 하고, 대리자를 지정하는 메서드를 만들어 보았습니다!!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/ProjectListViewController.swift:
```diff
@@ -0,0 +1,195 @@
+    @objc private func pressProjectCell(_ sender: UIGestureRecognizer) {
+        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        let location = sender.location(in: tableView)
+        guard let indexPath = tableView.indexPathForRow(at: location),
+              let project = dataSource?.itemIdentifier(for: indexPath)
+        else {
+            return
+        }
```
이게 더 자연스럽다면 요렇게 가셔도 됩니다! 
위에 댓글 남겨서 다시 코멘트했어요~!

junbangg commented on ProjectManager/ProjectManager/Sources/View/Controllers/ProjectListViewController.swift:
```diff
@@ -0,0 +1,195 @@
+            return
+        }
+
+        State.allCases.filter { $0 != self.state }.forEach { state in
```
연속으로 `.` 이 무려 3개가 있네요! 
살짝 다이어트를 시킬 방법은 없을까요?? 좋은 수가 없다면 
```
State.allCases.filter { $0 != self.state }
              .forEach { state in
```
요렇게라도?!

wonbi92 replied:
로직 자체는 좋은 방향성이라는 생각이 들어서 약간의 개행과 상수선언으로 다이어트를 약간 시켜보았습니다!

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/AddViewModel.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    weak var delegate: ViewModelDelegate?
+    
+    func addProject(title: String, deadline: Date, description: String) {
+        project = Project(title: title, deadline: deadline, description: description, state: .todo)
+    }
```
요렇게 하는 방법이랑
`Project` 를 밖에서 만들어서 주입 시키는 방법이랑 어떤 장단점/차이가 있을까요??

wonbi92 replied:
밖에서 주입시킨다면 매개변수의 타입을 추상화시켜서 좀 더 재사용성이 있는 메서드를 만들 수 있을것 같습니다!
저희가 요렇게 한 이유는 뷰컨트롤러가 모델도 모르는 상태를 만들어 보고 싶어서 요렇게 구현해 보았습니다!
알라딘은 어떤 것이 더 좋은 방향성이라고 생각하시나요??

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/EditViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    func updateProject(title: String, deadline: Date, description: String) {
+        project?.title = title
+        project?.deadline = deadline
+        project?.description = description
```
저~기 위에 댓글로 남겼던 `componentsHandler` 뷰요소로 쪼개는거 만약에 하시게 된다면
`Project`도 이 `ViewModel` 이 계속 들고 있을 필요도 없을? 수도 있겠네요

그냥 `Project` 처음 주입 받을때 분해하는 식으로? 하고

lxodud replied:
위에서 말씀드린것 처럼 id 값이 필요해서 Project를 가지고 있었는데 고민되었던 부분이 id만 따로 저장하고 있는 방법과 Project를 통째로 저장하는 방법이었습니다. 둘중에 어떤 방향성이 더 좋다고 생각하시는지 궁금합니다!!!

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/EditViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+        project?.deadline = deadline
+        project?.description = description
+        
+        delegate?.updateProject(project)
```
(위 코멘트에서 이어서)
여기서 `delegate` 한테 쏴야할때 새로운 `Project` 생성해서 보내는 방식은 안되는걸까요??

어차피 MainViewModel 에서 `id` 로 비교해서 수정하는거일것 같은데.. 요건 잘 확인해주십쇼🙏
의견일뿐!

wonbi92 replied:
위에서 코멘트 주신 부분부터 적용해서 리팩토링 해 보았습니다! 확실히 뷰가 더 깡통이 된거 같아요! 확장성이 더 좋아진것 같습니다 !! 👍

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/EditViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+        isEditing = state
+    }
+    
+    func setupProject(_ project: Project) {
```
그냥 `setProject` 로 해도 될듯하네요??? `up` 하는 코드는 딱히 없는것 같아서ㅎㅎ

lxodud replied:
set up은 뭔가 여러개를 설정하는 느낌이 강해서 set이 더 적절한 것 같습니다. 
저희 생각이 알라딘이 생각하신 부분이랑 동일한가요??

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/HeaderViewModel.swift:
```diff
@@ -0,0 +1,39 @@
+    var titleHandler: ((String?) -> Void)?
+    var cellCountHandler: ((String?) -> Void)?
+    
+    func setupTitle(_ title: String) {
```
`setTitle`?! ㅎㅎ

junbangg commented on ProjectManager/ProjectManager/Sources/ViewModel/MainViewModel.swift:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+    
+    private func changeProject(in list: inout [Project], to project: Project) {
+        guard let index = list.firstIndex(where: { $0.id == project.id }) else { return }
+        list[index] = project
+    }
```
아하 요기 있군요 비교해서 수정하는 코드가!

junbangg reviewed (changes requested):
수고하셨습니다!!

MVVM이랑 아키텍쳐 고민 많이 하시더니 뷰를 아주 깡통으로 만들어놓으셨군요..! 고민의 흔적이 아주 느껴졌습니다:)
MVVM 구조도 명확한 편이고 코드에 패턴이 보여서 가독성까지 좋아지는!!
아키텍쳐 설계가 잘되어서 그런게 아닐까 싶은데요!!🙌

코멘트 확인해주세요~
질문 주신 부분은 내일 달아두겠습니다~!

wonbi92 committed:
docs: README 수정

wonbi92 committed:
refactor: 메서드 및 프로퍼티 네이밍 수정

- 메서드 위치 변경

lxodud committed:
refactor: MainViewController의 프로퍼티를 외부에서 주입받게 변경

- ProjectListViewController인스턴스를 initializer로 주입받게 변경하였습니다.

wonbi92 committed:
refactor: EditViewController에서 Project타입을 바인딩하지 않고 각각의 프로퍼티 값을 바인딩하도록 리팩토링

- delegate은닉화 진행
- EditViewController와 AddViewController에서 옵셔널 바인딩을하는 로직을 뷰모델이 하도록 리팩토링
